TypeScriptPlugin/Implementation/ExportableFunction.swift

```diff
@@ -0,0 +1,97 @@
+// Copyright © 2024 Raycast. All rights reserved.
+
+import Foundation
+import SwiftSyntax
+
+struct ExportableFunction: @unchecked Sendable {
+  let attribute: String
+  let name: String
+  var parameters: [(name: String, type: TypeSyntax)] = []
+  let isAsync: Bool
+  let isThrowing: Bool
+  var returnType: TypeSyntax?
+}
+
+extension ExportableFunction {
+  var typescriptInterface: String {
+    var result = "async function \(name)("
+    result.append(parameters.map { "\($0.name): \(Self.typeScriptType(for: $0.type))" }.joined(separator: ", "))
+    result.append("): Promise<")
+    result.append(returnType.map { Self.typeScriptType(for: $0) } ?? "void")
+    result.append(">")
+    return result
+  }
+
+  func typescriptImplementation(runner: String) -> String {
+    var result = typescriptInterface
+    result.append(" {\n")
+    result.append(#"  return await \#(runner)("\#(name)""#)
+    if !parameters.isEmpty {
+      result.append(", ")
+      result.append(parameters.map(\.name).joined(separator: ", "))
+    }
+    result.append(")\n}")
+    return result
+  }
+
+  var javascriptInterface: String {
+    var result = "async function \(name)("
+    result.append(parameters.map(\.name).joined(separator: ", "))
+    result.append(")")
+    return result
+  }
+
+  func javascriptImplementation(runner: String) -> String {
+    var result = javascriptInterface
+    result.append(" {\n")
+    result.append(#"  return await \#(runner)("\#(name)""#)
+    if !parameters.isEmpty {
+      result.append(", ")
+      result.append(parameters.map(\.name).joined(separator: ", "))
+    }
+    result.append(")\n}")
+    return result
+  }
+}
+
+private extension ExportableFunction {
+  static func typeScriptType(for swiftType: TypeSyntax) -> String {
+    if let type = swiftType.as(OptionalTypeSyntax.self) {
+      return "\(typeScriptType(for: type.wrappedType)) | undefined"
+    } else if let type = swiftType.as(ImplicitlyUnwrappedOptionalTypeSyntax.self) {
+      return typeScriptType(for: type.wrappedType)
+    } else if let type = swiftType.as(SomeOrAnyTypeSyntax.self) {
+      return typeScriptType(for: type.constraint)
+    } else if case .some = swiftType.as(MetatypeTypeSyntax.self) {
+      return "any"
+    } else if let type = swiftType.as(ArrayTypeSyntax.self) {
+      return "\(typeScriptType(for: type.element))[]"
+    } else if let type = swiftType.as(DictionaryTypeSyntax.self) {
+      return "{ [key: \(typeScriptType(for: type.key))]: \(typeScriptType(for: type.value)) }"
+    }
+
+    guard let type = swiftType.as(IdentifierTypeSyntax.self),
+          case .identifier(let name) = type.name.tokenKind else { return "any" }
+
+    if let args = type.genericArgumentClause?.arguments {
+      switch name {
+      case "Set" where args.count == 1: return "Set<\(typeScriptType(for: args.first!.argument))>"
```

Comment:
Did you test this? I'd be surprised if that worked as a `Set` isn't JSON serialisable in JS

Reply:
I didn't test `Set`. I just ask ChatGPT whether Swift `Set` had an equivalent in TypeScript. @grzegorzkrukowski, would you mind testing this since you have everything end to end now.

Reply:
@dehesa @mathieudutour  caught it right when I have an argument being:

Set(3) { 'Barry Cade', 'Lorem Ipsum', 'Test Dummie' }  it stringified to just empty dictionary in JS '{}'

We will have to first make it an `Array.from(arg)` if we want to encode it, but not sure if it's worth an effort to complicate it or just keep it simple and support only Arrays.

Reply:
Actually after adding this to our implementation code:

```
if (arg instanceof Set) {
			commandArgs.push(JSON.stringify(Array.from(arg)));
		} else {
			commandArgs.push(JSON.stringify(arg));
		}
```

it all works :)

Reply:
I don't know how often is Set used, but it is probably not worth it

Reply:
On the other hand it's one-liner to make it work - just for sake of it ;)